### ISSUE TITLE
Program using Roslyn Compiler works in 4.8 but not in net5

### ISSUE THREAD

**Issue by jogibear9988** [dotnet/runtime#47059]
I've a program wich uses Roslyn to compile user code at runtime.

This Program uses some nugets wich are only netstandard1.4

The program works as expected in Net4.8
In netcore5 in need to add MetadataReferences to "netstandard" and "System.Runtime" via string to get it to work.
Is there another way?

See following Code:

    using Microsoft.CodeAnalysis;
    using Microsoft.CodeAnalysis.CSharp;
    using Microsoft.CodeAnalysis.Emit;
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Linq;
    using System.Reflection;

    namespace CompilerTest
    {
        class Program
        {
            static void Main(string[] args)
            {
                var code = @"
    using System;

    namespace CompilerTest
    {
        class Test
        {
            BetterWin32Errors.Win32Error bb;
        }
    }
    ";
            
                SyntaxTree syntaxTree = CSharpSyntaxTree.ParseText(code, new CSharpParseOptions(kind: SourceCodeKind.Regular, languageVersion: LanguageVersion.Latest));
                string assemblyName = Path.GetRandomFileName();

                var references = new List<MetadataReference>()
                {
                    MetadataReference.CreateFromFile(typeof(object).Assembly.Location),
                    MetadataReference.CreateFromFile(typeof(Enumerable).Assembly.Location),
                    MetadataReference.CreateFromFile(typeof(BetterWin32Errors.Win32Error).Assembly.Location),
                    //MetadataReference.CreateFromFile(Assembly.Load("netstandard").Location),
                    //MetadataReference.CreateFromFile(Assembly.Load("System.Runtime").Location)
                };

                var compilation = CSharpCompilation.Create(
                    assemblyName,
                    syntaxTrees: new[] { syntaxTree },
                    references: references,
                    options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary,
                        optimizationLevel: OptimizationLevel.Debug));

                using (var ms = new MemoryStream())
                {
                    EmitResult result = compilation.Emit(ms);

                    if (!result.Success)
                    {
                        IEnumerable<Diagnostic> failures = result.Diagnostics.Where(diagnostic =>
                            diagnostic.IsWarningAsError ||
                            diagnostic.Severity == DiagnosticSeverity.Error);

                        var compilerErrors = failures.Select(x => x.ToString());

                        Console.WriteLine("Compile Errors: " + compilerErrors);
                    }
                    else
                    {
                        Console.WriteLine("Compile Okay");
                    }
                    Console.ReadLine();
                }
            }
        }
    }

if you uncomment the lines:

                    //MetadataReference.CreateFromFile(Assembly.Load("netstandard").Location),
                    //MetadataReference.CreateFromFile(Assembly.Load("System.Runtime").Location)

it also works in netcore.

If not you get the Exception of the missing references.

**Comment by jogibear9988** [comment]
[CompilerTest.zip](https://github.com/dotnet/runtime/files/5822920/CompilerTest.zip)

**Comment by jogibear9988** [comment]
The Problem is here, how do I know wich of the Metadata references I need to add? Only by trial and error? Do I need to hardcode the names of them via string?

**Comment by jogibear9988** [comment]
BetterWin32Errors is only a sample, my real program uses many more nugets, I can not change all of them

**Comment by danmosemsft** [comment]
Does just netstandard alone work?

**Comment by jogibear9988** [comment]
what do you mean?
my project is net5, what should be only „netstandard“?

**Comment by danmosemsft** [comment]
What error message do you get, when only referencing netstandard?

**Comment by jogibear9988** [comment]
this one (sorry it's a german windows)

(8,31): error CS0012: Der Typ "Enum" ist in einer nicht referenzierten Assembly definiert. Fügen Sie einen Verweis auf die Assembly "System.Runtime, Version=4.0.20.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" hinzu.

but why do I also need to add netstandard? If I compile in vs, and add a reference to that lib, i also don't need to do more. it resolves the needed assemblies automaticaly. how could i achive this with roslyn embedded?

**Comment by danmosemsft** [comment]
@ericstj shouldnt type forwarding be avoiding this?

**Comment by ericstj** [comment]
Indeed, this is happening because of missing typeforwards. Those are provided in Sysytem.Runtime and netstandard in this case.  Those are needed in order to resolve the types in other assemblies (eg: the nuget packages).  It appears automatic in VS because MSBuild automatically references everything from the targeting pack when calling the compiler. You are calling the compiler yourself so you need to do what the build is doing.

**Comment by jogibear9988** [comment]
is there a way i can use the same logic as msbuild does while using roslyn compiler internal in my application?

**Comment by jogibear9988** [comment]
does thos mean I need the sdk version of netcore on my customers side, or is the runtime still enough?

**Comment by ericstj** [comment]
With respect to references MSBuild adds everything from the reference pack that is in `C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\<version>`.  I believe that you would need the SDK in order to have these files, though you could download them (they're also published as nuget packages).

This isn't dissimilar from what happens in .NETFramework.  In that case msbuild uses references from `C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\<version>` and it has similar rules for the typeforward assemblies in the `facades` subfolder.  Depending on the nuget package you might need to add references to these even on .NETFramework.

I took a closer look at this `BetterWin32Errors` nuget package and can explain why you are seeing differences.
```
betterwin32errors.0.2.0\lib\net20\BetterWin32Errors.dll
betterwin32errors.0.2.0\lib\net40\BetterWin32Errors.dll
betterwin32errors.0.2.0\lib\net45\BetterWin32Errors.dll
betterwin32errors.0.2.0\lib\netstandard1.4\BetterWin32Errors.dll
```

So this package contains a .NETFramework specific assembly in the `net45` folder.  That will build against `mscorlib` so all the types referenced will be present with a reference to `msorlib`.  On .NETCore the `netstandard1.4` assembly will be used, which requires additional references to resolve the types.  This explains the reason you are observing a difference.

Since you are asking about best-practices around using roslyn and resolving references I'm going to transfer this to the roslyn repository.  Perhaps they provide a way to resolve references dynamically and you could load them from the runtime directory as they are needed?
